File: python/oneibl/tests/test_one.py
import unittest
import numpy as np
from oneibl.one import ONE


class TestLoad(unittest.TestCase):

    def setUp(self):
        # Init connection to the database
        one = ONE(base_url='https://test.alyx.internationalbrainlab.org', username='test_user',
                  password='TapetesBloc18')
        eids = ['cf264653-2deb-44cb-aa84-89b82507028a', '4e0b3320-47b7-416e-b842-c34dc9004cf8']
        self.eid = eids[0]
        self.eid2 = eids[1]
        self.One = one

    def test_list(self):
        one = self.One
        # tests with a single input and a list input
        EIDs = [self.eid,
                [self.eid, self.eid2]]
        for eid in EIDs:
            dt = one.list(eid)  # returns dataset-type
            dt = one.list(eid, details=True)
            dt = one.list(eid, keyword='dataset-type')  # returns list
            dt = one.list(eid, keyword='dataset-type', details=True)  # returns SessionDataInfo
            for key in ('subject', 'users', 'lab', 'type', 'start_time', 'end_time'):
                dt = one.list(eid, keyword=key)  # returns dataset-type
                print(key, ': ', dt)
            ses = one.list(eid, keyword='all')
            usr, ses = one.list(eid, keyword='users', details=True)

    def test_list_error(self):
        one = self.One
        a = 0
        eid = self.eid
        try:
            one.list(eid, keyword='tutu')  # throws an error
        except KeyError:
            a = 1
            pass
        self.assertTrue(a == 1)

    def test_load(self):
        # Test with 3 actual datasets predefined
        one = self.One
        dataset_types = ['clusters.peakChannel', 'clusters._phy_annotation', 'clusters.probes']
        eid = ('https://test.alyx.internationalbrainlab.org/'
               'sessions/' + self.eid)
        t, cr, cl = one.load(eid, dataset_types=dataset_types)
        d = one.load(eid, dataset_types=dataset_types, dclass_output=True)
        self.assertTrue(np.all(d.data[0] == t))
        # Now load with another dset inbetween that doesn't exist
        t_, cr_, cl_ = one.load(eid, dataset_types=['clusters.peakChannel', 'turlu',
                                                    'clusters.probes'])
        self.assertTrue(np.all(t == t_))
        self.assertTrue(np.all(cl == cl_))
        self.assertTrue(cr_ is None)

    def test_load_empty(self):
        # Test with a session that doesn't have any dataset on the Flat Iron
        one = self.One
        eid = self.eid
        dataset_types = ['wheel.velocity', 'wheel.timestamps']
        a, b = one.load(eid, dataset_types=dataset_types)
        self.assertTrue(a is None and b is None)

    def test_load_from_uuid(self):
        # Test the query with only the UUID string and not the full URL (no data here)
        one = self.One
        eid = self.eid
        dataset_types = ['wheel.velocity', 'wheel.timestamps']
        aa = one.load(eid, dataset_types=dataset_types)
        self.assertTrue(len(aa) == 2)

    def test_load_str(self):
        one = self.One
        eid = self.eid
        a = one.load(eid, 'eye.raw')
        self.assertTrue(len(a) == 1)

    def test_load_all_data_available(self):
        # Test without a dataset list should download everything and output a dictionary
        one = self.One
        eid = self.eid2
        a = one.load(eid)
        self.assertTrue(len(a.data) == 5)

    def test_search_simple(self):
        one = self.One
        # Test users
        usr = ['olivier', 'nbonacchi']
        sl, sd = one.search(users=usr, details=True)
        self.assertTrue(isinstance(sl, list) and isinstance(sd, list))
        self.assertTrue(all([set(usr).issubset(set(u)) for u in [s['users'] for s in sd]]))
        # when the user is a string instead of a list
        sl1, sd1 = one.search(users=['olivier'], details=True)
        sl2, sd2 = one.search(users='olivier', details=True)
        self.assertTrue(sl1 == sl2 and sd1 == sd2)
        # test for the dataset type
        dtyp = ['spikes.times', 'titi.tata']
        sl, sd = one.search(dataset_types=dtyp, details=True)
        self.assertTrue(len(sl) == 0)
        dtyp = ['channels.site']
        sl, sd = one.search(dataset_types=dtyp, details=True)
        self.assertTrue(len(sl) == 2)
        dtyp = ['spikes.times', 'channels.site']
        sl, sd = one.search(dataset_types=dtyp, details=True)
        self.assertTrue(len(sl) == 1)
        # test empty return for non-existent user
        self.assertTrue(len(one.search(users='asdfa')) == 0)
        # test search with the lab keyword
        self.assertTrue(len(one.search(lab='zadorlab')) == 1)

    def test_session_does_not_exist(self):
        eid = 'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee'
        self.assertRaises(FileNotFoundError, self.One.load, eid)


if __name__ == '__main__':
    unittest.main()

File: matlab/tests/test_one.m
classdef test_one < matlab.unittest.TestCase
 
    properties
        one
        eid
        eid2
    end
 
    methods(TestMethodSetup)
        function createObject(testCase)
            testCase.one = One('alyx_login', 'test_user', 'alyx_pwd', 'TapetesBloc18',...
                'alyx_url', 'https://test.alyx.internationalbrainlab.org');
            testCase.eid = 'cf264653-2deb-44cb-aa84-89b82507028a';
            testCase.eid2 = '4e0b3320-47b7-416e-b842-c34dc9004cf8';
        end
    end
 
    methods(Test)
 
        function test_list(testCase)
            eid_ = testCase.eid;
            eids_ = {testCase.eid, testCase.eid2};
            %% first functionality is a straight list of datasets without keyword
            % test for a single session, list datasets
            [dt, details]= testCase.one.list(eid_);
            testCase.verifyEqual(length(dt),29);
            % test for multiple sessions, list datasets
            [dt, details]= testCase.one.list(eids_);
            testCase.verifyEqual(length(dt),29);
            testCase.verifyEqual(length(details.url),34);
            %% second functionality is a straight query to the endpoint
            [list, ~] = testCase.one.list([],'keyword','data');
            testCase.assertTrue(length(list)>40);
            testCase.one.list([],'keyword','subject');
            testCase.one.list([],'keyword','users');
            lab_list = testCase.one.list([],'keyword','labs'); 
            testCase.assertTrue(length(lab_list) >=3);
            %% third functionality is a query on the session table attributes:
            for key = {'subject', 'users', 'lab', 'type', 'start_time', 'end_time'}
                dt = testCase.one.list(eids_, 'keyword', key);
                % this tests that the function aggregates unique sets
                if any(strcmp(key, {'type'}))
                    testCase.assertTrue(length(dt)==1)
                else
                    testCase.assertTrue(length(dt)==2)
                end
            end
            %% This returns a structure of arrays for each session with all fields
            ses = testCase.one.list(eids_, 'keyword', 'all');
            testCase.assertTrue(all(structfun(@length, ses)==2))
        end
% 
%     def test_list_error(self):
%         one = self.One
%         a = 0
%         eid = '86e27228-8708-48d8-96ed-9aa61ab951db'
%         try:
%             one.list(eid, keyword='tutu')  # throws an error
%         except KeyError:
%             a = 1
%             pass
%         self.assertTrue(a == 1)
        
        function test_search(testCase)
            % test datenum string and double
            eids  = testCase.one.search('users','olivier','date_range',datenum(['2018-08-24'; '2018-08-24']));
            eids_ = testCase.one.search('users','olivier','date_range',(['2018-08-24'; '2018-08-24']));
            testCase.verifyEqual(eids , eids_, testCase.eid)
            % test empty
            eids = testCase.one.search('users','etsitunexistaispas');
            testCase.assertTrue(isempty(eids));
            % test 2 users and single session output
            [eids, ses]= testCase.one.search('users',{'olivier', 'nbonacchi'});
            % test single user string and multiple sessions output
            [eids, ses]= testCase.one.search('users','olivier');
            % test for the dataset type
            dtyp = {'spikes.times', 'titi.tata'};
            [eids, ses]= testCase.one.search('dataset_types',dtyp);
            testCase.assertTrue(isempty(eids))
            dtyp = 'channels.site';
            [eids, ses]= testCase.one.search('dataset_types',dtyp);
            testCase.assertTrue(length(eids)==2)
            % test automatic typo correction
            [eids_]= testCase.one.search('data',dtyp);
            testCase.assertEqual(eids, eids_);
            dtyp = {'spikes.times', 'channels.site'};
            [eids, ses] = testCase.one.search('dataset_types',dtyp);
            % test empty return for non-existent user
            [eids, ses] = testCase.one.search('subjects',{'turlu'});
        end
        
        function test_load(testCase)
            % Test with 3 actual datasets predefined
            dataset_types = {'clusters.peakChannel', 'clusters._phy_annotation', 'clusters.probes'};
            eid_ = ['https://test.alyx.internationalbrainlab.org/sessions/' testCase.eid];
            [pc, pa, cp] = testCase.one.load(eid_, 'dataset_types', dataset_types);
            % same with dclass output and check correspondency
            D = testCase.one.load(eid_, 'dataset_types', dataset_types, 'dclass_output', true);
            testCase.assertTrue(all(pc==D.data{1}) & all(pa==D.data{2}) & all(cp==D.data{3}))
            % Test with a session that doesn't have any dataset on the Flat Iron
            dataset_types = {'wheel.velocity', 'wheel.timestamps'};
            [a1, a2] = testCase.one.load(eid_, 'dataset_types', dataset_types);
            testCase.assertTrue(isempty(a1) & isempty(a2))
            % Test with an empty dataset interleaved
            dataset_types = {'clusters.peakChannel', 'turlu', 'clusters.probes'};
            [pc_, tu, cp_] = testCase.one.load(eid_, 'dataset_types', dataset_types);
            testCase.assertTrue(all(pc==pc_) & isempty(tu) & all(cp==cp_))
            % test a single dtype in scalar
            a = testCase.one.load(testCase.eid, 'dataset_types', 'eye.area');
            % Test without a dataset list should download everything and output a dictionary
            a = testCase.one.load(testCase.eid2);
            testCase.assertTrue(length(a.data) == 5)
        end
        
    end
end